Quickstart docs page (#654)
POSYDON-code/POSYDON: docs/_source/getting-started/first-grids.rst

.. _first-grids:

Looking at our POSYDON grids
============================

Installation and Data Download
------------------------------

> First, make sure POSYDON has been installed on a conda environment. If you have not yet completed this step, please take a look at our installation guide `here <installation-guide>`_.

Reviewer: ```suggestion
First, make sure POSYDON has been installed in a ``conda`` environment. If you have not yet completed this step, please take a look at our installation guide `here <installation-guide>`_.
```

.. code-block:: bash

    get-posydon-data DR2_1Zsun

Loading a grid
--------------

You can access our grids of MESA simulations using the ``PSyGrid`` module. To load a grid, first import the module, instantiate a new ``PSyGrid`` object, and pass the argument ``filepath`` to open a grid. In this example we load the HMS-HMS grid at solar metallicity:

.. code-block:: python

    from posydon.grids.psygrid import PSyGrid
    PATH_TO_POSYDON_DATA = '/YOUR/POSYDON_DATA/PATH'
    grid_path = f'{PATH_TO_POSYDON_DATA}/POSYDON_data/HMS-HMS/1e+00_Zsun.h5'
    mygrid = PSyGrid(filepath=grid_path)

Grid contents
-------------
You can check the contents of the ``PSyGrid`` object with a print command: 

.. code-block:: python

    print(mygrid)

This will provide a summary of grid metadata including how many runs are in the grid and what information is stored about them. 

You can get a list of the :ref:`grid configuration properties <tab_grid_properties>` available for your :samp:`PSyGrid` object with:

.. code-block:: python

    mygrid.config.keys()

You can access the value of any grid configuration property "PROPERTY" with 

> :samp:`mygrid.config[{PROPERTY}]`.

Reviewer: ```suggestion
:samp:`mygrid.config[{PROPERTY}]`. Viewing this information can be useful if you are curious about how the grid was run and which options were included when postprocessing the grid (this is an advanced topic that you can read more about :ref:`here </components-overview/post_processing/pipeline.rst>`).
```

> Next, you can look at the initial and final values of the runs. All the values

Reviewer: ```suggestion
To maintain a relatively low data storage footprint, the PSyGrid object does not store all of the raw MESA stellar evolution data, but you can use it to look at the initial and final values and downsampled histories of the MESA simulations. All the values
```

are available at :samp:`mygrid.initial_values` and :samp:`mygrid.final_values`,

> respectively. To get a tuple of all the available values use

Reviewer: ```suggestion
respectively. To get a tuple of all the available column names use
```

.. code-block:: python

    mygrid.initial_values.dtype.names
    mygrid.final_values.dtype.names

You can access the initial value of any physical grid property "PHYS" with 
:samp:`mygrid.initial_values[{PHYS}]`. It will return a numpy array with the 

> values of this property for all the runs. 

Reviewer: ```suggestion
initial values of this property for all the MESA runs in this grid. 
```

Note that these physical properties of the binaries in the grid are different 
from the grid configuration properties listed above. 

> Then, you can find the initial mass of star 1 in the third MESA run with

Reviewer: ```suggestion

As an example, you can find the initial mass of the primary star (``star_1_mass``) in one of the grid's MESA binary star simulations with
```

.. code-block:: python

    mygrid.initial_values['star_1_mass'][2]

.. note::

>     Remember that the first run has the index :samp:`0` and the last one

Reviewer: ```suggestion
    In the example, we used an arbitrary index `2` to select a binary, but the indices run from :samp:`0` to 
```

    :samp:`len(mygrid)-1`.

You can retrieve individual runs by index. :samp:`mygrid[{IDX}]` is a

> :samp:`PSyRunView` object, which contains the data of the run of index 

Reviewer: ```suggestion
:samp:`PSyRunView` object which contains data pertaining to that particular run of index 
```

> :samp:`IDX`. The :samp:`PSyRunView` object contains seven components:

Reviewer: ```suggestion
:samp:`i`. The :samp:`PSyRunView` object contains seven components:
```

.. table:: :samp:`PSyRunView` object components

    ================  ===========
    Component         Description
    ================  ===========

>     'initial_values'  all initial values of the run

Reviewer: ```suggestion
    'initial_values'  all initial (i.e., at the first time step of the MESA simulation) values of the run
```

    'final_values'    all final values of the run including termination flags
    'binary_history'  the binary history

>     'history1'        the history of star 1

Reviewer: ```suggestion
    'history1'        the downsampled evolution data of star 1
```

>     'history2'        the history of star 2

Reviewer: ```suggestion
    'history2'        the downsampled evolution data of star 2
```

>     'final_profile1'  the final profile of star 1

Reviewer: ```suggestion
    'final_profile1'  the profile (properties vs. mass or radius) of star 1 at the final timestep
```

>     'final_profile2'  the final profile of star 2

Reviewer: ```suggestion
    'final_profile2'  the profile of star 2 at the final timestep
```

    ================  ===========

> Again, you can check for the contents of the individual runs with

Reviewer: ```suggestion
Again, you can check for the valid column names of the individual runs with
```

:samp:`dtype.names`, e.g.

.. code-block:: python

    myrun = mygrid[0]
    myrun['binary_history'].dtype.names

> The example above finds the initial mass of star 1 in the third MESA run by 

Reviewer: ```suggestion
Before, we retrieved the initial mass of star 1 in the binary indexed at ` 2` by
```

indexing the list :samp:`mygrid.initial_values`. 
You can get the same value from the list of initial values associated with a 

> single MESA run: 

Reviewer: ```suggestion
single MESA run using the `PSyRunView` object via: 
```
